stm32f303 - irq conflict between CAN and USB

I'm porting zephyr over to a custom board based on a stm32f303rc.
I'm trying to use both of CAN and USB.
If you enable both and build you get an error (see later).

Looking at STM32 datasheet - ( en.DM00043574.pdf)  at page 286 you can see that both position 19 and 20 are shared between USB and CAN but footnote 1 says

> It is possible to remap the USB interrupts (USB_HP, USB_LP and USB_WKUP) on interrupt lines 74, 75 and 76
> respectively by setting the USB_IT_RMP bit in the Section 12.1.1: SYSCFG configuration register 1 (SYSCFG_CFGR1) on
> page 245.

The error

> FAILED: zephyr/isr_tables.c
> cd /home/giellamo/projects/swhard/hcb/build/zephyr && /opt/toolchains/zephyr-sdk/arm-zephyr-eabi/bin/arm-zephyr-eabi-objcopy --input-target=elf32-littlearm --output-target=binary
>  --only-section=.intList /home/giellamo/projects/swhard/hcb/build/zephyr/zephyr_prebuilt.elf isrList.bin && /usr/bin/python3 /home/giellamo/projects/zephyrproject/zephyr/arch/com
> mon/gen_isr_tables.py --output-source isr_tables.c --kernel /home/giellamo/projects/swhard/hcb/build/zephyr/zephyr_prebuilt.elf --intlist isrList.bin   --sw-isr-table --vector-ta
> ble
> gen_isr_tables.py: multiple registrations at table_index 20 for irq 20 (0x14)
> Has IRQ_CONNECT or IRQ_DIRECT_CONNECT accidentally been invoked on the same irq multiple times?
> Traceback (most recent call last):
>   File "/home/giellamo/projects/zephyrproject/zephyr/arch/common/gen_isr_tables.py", line 316, in <module>
>     main()
>   File "/home/giellamo/projects/zephyrproject/zephyr/arch/common/gen_isr_tables.py", line 307, in main
>     + "\nHas IRQ_CONNECT or IRQ_DIRECT_CONNECT accidentally been invoked on the same irq multiple times?"
>   File "/home/giellamo/projects/zephyrproject/zephyr/arch/common/gen_isr_tables.py", line 36, in error
>     raise Exception()
> Exception
> ninja: build stopped: subcommand failed.


**To Reproduce**
Steps to reproduce the behavior:
1. mkdir build; cd build
2. west build -t menuconfig -> enable both USB and CAN
3. west build -b hcb -- -DBOARD_ROOT=/home/giellamo/projects/swhard/hcb
3. make
4. See error

**Expected behavior**
It should be possibile to relocate interrupts

**Impact**
Severe cause I need both of the peripherals

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain  Zephyr SDK
 - Commit SHA: 05ced0e79c455356f5cb921eabb2b32f79756b95

--- a/drivers/clock_control/clock_stm32f0_f3.c
+++ b/drivers/clock_control/clock_stm32f0_f3.c
@@ -77,6 +77,13 @@ void config_enable_default_clocks(void)
 	/* Enable System Configuration Controller clock. */
 	LL_APB1_GRP2_EnableClock(LL_APB1_GRP2_PERIPH_SYSCFG);
 #endif
+#else
+#if defined(CONFIG_USB_DC_STM32) && defined(SYSCFG_CFGR1_USB_IT_RMP)
+	/* Enable System Configuration Controller clock. */
+	/* SYSCFG is required to remap IRQ to avoid conflicts with CAN */
+	/* cf §14.1.3, RM0316 */
+	LL_APB2_GRP1_EnableClock(LL_APB2_GRP1_PERIPH_SYSCFG);
+#endif
 #endif /* !CONFIG_SOC_SERIES_STM32F3X */
 }
 

--- a/drivers/usb/device/usb_dc_stm32.c
+++ b/drivers/usb/device/usb_dc_stm32.c
@@ -386,6 +386,20 @@ int usb_dc_attach(void)
 
 	LOG_DBG("");
 
+#ifdef SYSCFG_CFGR1_USB_IT_RMP
+	/*
+	 * STM32F302/F303: USB IRQ collides with CAN_1 IRQ (§14.1.3, RM0316)
+	 * Remap IRQ by default to enable use of both IPs simultaneoulsy
+	 * This should be done before calling any HAL function
+	 */
+	if (LL_APB2_GRP1_IsEnabledClock(LL_APB2_GRP1_PERIPH_SYSCFG)) {
+		LL_SYSCFG_EnableRemapIT_USB();
+	} else {
+		LOG_ERR("System Configuration Controller clock is "
+			"disabled. Unable to enable IRQ remapping.");
+	}
+#endif
+
 	/*
 	 * For STM32F0 series SoCs on QFN28 and TSSOP20 packages enable PIN
 	 * pair PA11/12 mapped instead of PA9/10 (e.g. stm32f070x6)
@@ -395,7 +409,7 @@ int usb_dc_attach(void)
 		LL_SYSCFG_EnablePinRemap();
 	} else {
 		LOG_ERR("System Configuration Controller clock is "
-			"disable. Unable to enable pin remapping."
+			"disabled. Unable to enable pin remapping.");
 	}
 #endif
 

--- a/dts/arm/st/f3/stm32f302.dtsi
+++ b/dts/arm/st/f3/stm32f302.dtsi
@@ -8,6 +8,11 @@
 
 / {
 	soc {
+		usb: usb@40005c00 {
+			/* Remap USB_LP IRQ to enable use with CAN_1 */
+			interrupts = <75 0>;
+		};
+
 		i2c2: i2c@40005800 {
 			compatible = "st,stm32-i2c-v2";
 			clock-frequency = <I2C_BITRATE_STANDARD>;

--- a/dts/arm/st/f3/stm32f303.dtsi
+++ b/dts/arm/st/f3/stm32f303.dtsi
@@ -8,6 +8,11 @@
 
 / {
 	soc {
+		usb: usb@40005c00 {
+			/* Remap USB_LP IRQ to enable use with CAN_1 */
+			interrupts = <75 0>;
+		};
+
 		i2c2: i2c@40005800 {
 			compatible = "st,stm32-i2c-v2";
 			clock-frequency = <I2C_BITRATE_STANDARD>;

